From 4f25303bd3957efcf563200cc2b473636f93bd69 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:14:43 -0400
Subject: [alsa] oxygen: add the separate SPI waiting function

Message-id: <140972823573-47-patchset-jkysela@redhat.com>
Patchwork-id: 90747
O-Subject: [RHEL 7.1 PATCH 046/442] ALSA: oxygen: add the separate SPI waiting function
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

The oxygen_wait_spi() function now performs waiting when the
SPI bus completes a transaction. Introduce the timeout error
checking and increase timeout to 200 from 40.

Signed-off-by: Roman Volkov <v1ron@mail.ru>
Signed-off-by: Clemens Ladisch <clemens@ladisch.de>

(picked from kernel: 10dd44dc88419ba5278435669ac9e35f9f410c7e)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/oxygen/oxygen_io.c b/sound/pci/oxygen/oxygen_io.c
index 521eae4..4b88c7a 100644
--- a/sound/pci/oxygen/oxygen_io.c
+++ b/sound/pci/oxygen/oxygen_io.c
@@ -194,6 +194,24 @@ void oxygen_write_ac97_masked(struct oxygen *chip, unsigned int codec,
 }
 EXPORT_SYMBOL(oxygen_write_ac97_masked);
 
+static int oxygen_wait_spi(struct oxygen *chip)
+{
+ unsigned int count;
+
+ /*
+  * Higher timeout to be sure: 200 us;
+  * actual transaction should not need more than 40 us.
+  */
+ for (count = 50; count > 0; count--) {
+  udelay(4);
+  if ((oxygen_read8(chip, OXYGEN_SPI_CONTROL) &
+      OXYGEN_SPI_BUSY) == 0)
+   return 0;
+ }
+ snd_printk(KERN_ERR "oxygen: SPI wait timeout\n");
+ return -EIO;
+}
+
 void oxygen_write_spi(struct oxygen *chip, u8 control, unsigned int data)
 {
  unsigned int count;
-- 
1.7.1